### PR TITLE
Stream support for BLAS Handles.

### PR DESCRIPTION
**Pull request by colesbury**
I've rebased and included Jeff's changes from the previous PR.

maskedCopy implemented
generic Reduce kernels

### CODE REVIEW

**Comment by colesbury** [comment]
@dominikgrewe I believe this includes all the requested changes on #167

**Comment by colesbury** [comment]
(Remove THCDeviceTensor* files)

**Comment by soumith** [comment]
hey @dominikgrewe all your issues were addressed in this PR. Does this look okay to you?

File: lib/THC/THCTensorMasked.cu [hunk 1]
```diff
@@ -44,11 +40,85 @@ void THCudaTensor_maskedFill(THCState* state,
   THCudaCheck(cudaGetLastError());
 }
 
+struct TensorMaskedCopyOp {
+  TensorMaskedCopyOp(float* s, float* bm)
+      : src(s),
+        baseMask(bm) {
```
Comment (dominikgrewe) [inline lib/THC/THCTensorMasked.cu]: `baseMask` isn't used.

**Comment by dominikgrewe** [comment]
Sorry for the delay. I was away for a week.

I just have a small comment, otherwise it looks good to me. Thanks guys!

**Comment by colesbury** [comment]
I've removed the unused "baseMask"

**Pull request merged**